src/olympia/versions/tasks.py

```diff
@@ -126,12 +156,30 @@ def extract_version_to_git(version_id, author_id=None, note=None):
     log.info('Extracting {version_id} into git backend'.format(
         version_id=version_id))
 
-    repo = AddonGitRepository.extract_and_commit_from_version(
-        version=version, author=author, note=note)
-
-    log.info('Extracted {version} into {git_path}'.format(
-        version=version_id, git_path=repo.git_repository_path))
+    try:
+        repo = AddonGitRepository.extract_and_commit_from_version(
+            version=version, author=author, note=note)
 
+        log.info('Extracted {version} into {git_path}'.format(
+            version=version_id, git_path=repo.git_repository_path))
+    except BrokenRefError as err:
+        # This is needed to prevent a potential infinite loop if a broken
+        # reference is detected in `extract_addon_to_git()`, which we call
+        # later in this block.
+        if stop_on_broken_ref:
+            raise err
+
+        addon = version.addon
+        # Reset the git hash of each version of the add-on related to the
+        # current add-on because we want to re-extract everything.
+        versions = addon.versions.update(git_hash='')
+        # Retrieve the repo for the add-on and delete it.
+        addon_repo = AddonGitRepository(addon, package_type='addon')
+        addon_repo.delete()
+        log.warn('Deleted the git addon repository for addon_id={} because we '
+                 'detected a broken ref.'.format(addon.id))
+        # Create a task to re-extract the add-on.
+        extract_addon_to_git.delay(addon.pk)
```

Comment:
Might need something to avoid duplicate tasks for the same add-on. Also, before deleting the repos, you might want to add something to make sure other version extraction tasks for the same add-on won't be processed.

Reply:
Yep, those are the two questions I had in mind and that's why I opened this draft now.

> Might need something to avoid duplicate tasks for the same add-on.

This can probably be solved with some sort of _lock_.

> Also, before deleting the repos, you might want to add something to make sure other version extraction tasks for the same add-on won't be processed.

I am not sure to know how to do that yet. The other problematic scenario would be:

1. broken ref, repo deletion
2. task to re-extract add-on starts
3. new version is submitted by developer

How do we make sure this last version gets extracted?

I think this is where using rabbitmq acknowledgment would be nice because we could put the task back into the queue if its related add-on is being re-extracted. But maybe we have something to solve this problem already.

Reply:
Something something https://docs.celeryproject.org/en/latest/faq.html#faq-acks-late-vs-retry combined with some sort of lock might help.

I don't think we anything to retry skipped/failed extraction of a specific version at the moment. We do know about them because they are missing `source_git_hash` or `git_hash` on the `Version` instance, but I don't think we currently try to catch up later and re-extract.

Reply:
Do you think it would be practical to add a new field on `Addon` (`git_extraction_in_progress` but shorter) that we could us:

- in `extract_addon_to_git()` to avoid this task to be executed more than once
- in `extract_version_to_git()` to... re-submit itself?

That would probably solve the issues we're discussing here. Maybe we could add a small delay before re-submitting the `extract_version_to_git` task from within `extract_version_to_git`.

WDYT?

Reply:
That could work, yeah. I would use a separate model with a `models.OneToOneField(Addon, primary_key=True, on_delete=models.CASCADE)` instead of adding a new field to `Addon`, though.